moped-editor/src/views/projects/projectView/KnackSync.js

```diff
@@ -0,0 +1,217 @@
+        })
+          .then(response => response.json())
+          .then(
+            result => {
+              if (result.errors) {
+                // knack error
+                snackbarHandler({
+                  severity: "warning",
+                  message: "Success: Data Tracker sync failed.",
+                });
+                return Promise.reject(result);
+              } else {
+                // get-state success
+                project.moped_project[0].currentKnackState = result;
+                return fetch(buildUrl(), {
+                  method: getHttpMethod(),
+                  headers: buildHeaders(),
+                  body: buildBody(),
+                });
+              }
+            },
+            error => {
+              // get-state fetch error
+              return Promise.reject(error);
+            }
+          )
+          .then(response => response.json())
+          .then(
+            result => {
+              if (result.errors) {
+                // knack error
+                snackbarHandler({
+                  severity: "warning",
+                  message: "Success: Data Tracker sync failed.",
+                });
+              } else {
+                // successful update
+                snackbarHandler({
+                  severity: "success",
+                  message: "Success: Project data pushed to Data Tracker",
+                });
+              }
+            },
+            error => {
+              // fetch error
+              snackbarHandler({
+                severity: "warning",
+                message: "Success: Data Tracker sync failed.",
+              });
+            }
+          );
+      } else {
+        // creating new knack record
+        project.moped_project[0].currentKnackState = {};
+        fetch(buildUrl(), {
+          method: getHttpMethod(),
+          headers: buildHeaders(),
+          body: buildBody(),
+        })
+          .then(response => response.json())
+          .then(
+            result => {
+              if (result.errors) {
+                // knack error
+                snackbarHandler({
+                  severity: "warning",
+                  message: "Success: Data Tracker sync failed.",
+                });
+                return Promise.reject(result);
+              } else {
+                // knack fetch success
+                return Promise.resolve(result);
+              }
+            },
+            error => {
+              // fetch error
+              snackbarHandler({
+                severity: "warning",
+                message: "Success: Data Tracker sync failed.",
+              });
+              return Promise.reject(error);
+            }
+          )
+          .then(knack_record => {
+            mutateProjectKnackId({
+              variables: {
+                project_id: project.moped_project[0].project_id,
+                knack_id: knack_record.record.id,
+              },
+            });
```

Comment:
Shouldn't this also be part of the chain? What would happen if this mutation fails?

Reply:
related to above, when you chain these together i believe you can use a single `.catch()` at the end of the chain that will catch errors at any stage.

that said, i'm seeing where Sergio was going suggesting a hook. it would be quite bad if a knack project where created but the mutation failed. it would add quite a bit of state to isolate the two separate calls with hooks, but that might make it easier to build in some resiliency. if we want to go that route i think it's worthy of a separate PR.

Reply:
fwiw [here's](https://github.com/cityofaustin/atd-moped/blob/3f5539935dec8a93dd7163ff69d0adb9ff2037a5/moped-editor/src/utils/socrataHelpers.js#L8) an example of a hook-ified fetch.

Reply:
It should for sure - thank you for pointing out that this wasn't handled correctly.